### ISSUE TITLE
Does not load basic programs properly

### ISSUE THREAD

**Issue by JesperJuul** [vc64web/virtualc64web#188]
Great work!
As far as I can tell, the emulator does not work with BASIC .prg programs. They are loaded, but crash soon. LIST shows a garbled program. The same program will work if packaged in a .d64 file.

I suspect this is because the end of the BASIC program (location 45/46) is not set properly when loading .prg files. (Hence variables override the program.)

Thank you

**Comment by mithrendal** [comment]
Hi 🙋🏻‍♀️ @JesperJuul 

Can you attach such a .prg file ? I will try it with the current virtual C64 version 4.6 for mac whether it is already solved there. The web version is still at version 4.4. 

Thank you for reporting

**Comment by JesperJuul** [comment]
Sure! Here goes. It has been the same for any BASIC .prg I have tried.

[soft scroller v1.zip](https://github.com/vc64web/virtualc64web/files/11938945/soft.scroller.v1.zip)

**Comment by mithrendal** [comment]
reproduced in latest 4.6 build

**Comment by dirkwhoffmann** [comment]
> I suspect this is because the end of the BASIC program (location 45/46) is not set properly when loading .prg files. (Hence variables override the program.)

Yes, this is correct. At the moment, VirtualC64 does not modify any other memory locations than the ones that are read in from the file. Dragging in a PRG works in VICE, but I think they don't implement real flashing like in VirtualC64. After dragging in the file, there is a disk present in drive 8:

<img width="832" alt="Bildschirmfoto 2023-07-04 um 15 47 09" src="https://github.com/vc64web/virtualc64web/assets/12561945/1999909e-706c-4556-aade-c4906f8548a6">

Bottom line: Right now, BASIC programs need to take the D64 route in VirtualC64. In the Mac version, the user can choose if he wants a PRG to be flashed or converted to a floppy disk. Maybe @mithrendal can adapt something similar in his web emulator.

**Comment by mithrendal** [comment]
@dirkwhoffmann @JesperJuul 
ok ... then we could bring up a dialog when user drops a .prg file in case when at the same time also a drive is connected? ... the dialog could have two buttons `flash` and `insert into drive8` ... when user choses `insert into drive8` we might additionally also ask for optional auto typing of load and run commands. Would that be fine ?

what also comes to my mind is that when browsing the csdb which contain a lot of demo `.prg` files which mostly are NOT basic then this asking about flashing/inserting would introduce an additional dialog step ... 

It would be cool when the emulator could detect somehow whether it is  a .prg file which  contains only basic ... _(in which case it should always auto insert the prg file as a disk without asking for optional flashing because flashing would always be broken for basic .prgs anyways. Commodore should really have defined another extension like .bas for example or am I wrong ?)_ And when it is not a basic .prg maybe we just would leave it as it is now i.e. flashing without asking the user. Then the user would not need to make a choice ... 

I found some tools for converting bas to prg and prg to bas, which also describe the layout of a c64 basic prg file
https://github.com/abbrev/prg-tools

Is it maybe possible to detect a "basic" prg with the knowledge of that described byte layout ?

**Comment by JesperJuul** [comment]
I haven't looked at the code, but isn't there a way to just set 45/46 to the last loaded byte after having loaded a .prg file? The appears to be much simpler.

**Comment by JesperJuul** [comment]
I am pretty sure it's just 45/46. When you load a program, locations 45/46 are set to the last byte loaded+1. to make sure variables don't override the program. It's the same regardless of whether the program is in BASIC or machine code.

**Comment by mithrendal** [comment]
ok then I will add these three lines to the proposed location in c64.cpp

```cpp
    auto end_of_prg=(u64)(mem.ram + addr + size + 1);
    mem.poke(45, end_of_prg & 0xff);
    mem.poke(46, end_of_prg>>8);
```

lets see whether it works 🙉or whether it will crash ...

dragging the scroller prg into vc64web now... 

<img width="994" alt="image" src="https://github.com/vc64web/virtualc64web/assets/17108995/9eca9191-2256-40e1-a1ac-741a028588ab">

@JesperJuul @dirkwhoffmann  you both are super cool ... it really works !! ✌️

**Comment by mithrendal** [comment]
I just looked at the values and added together what I saw there available to add 😬

so I first added these values by guessing...

```cpp
auto end_of_prg=(u64)(mem.ram + addr + size + 1);
```

I thought that made sense but resulted in these above different values in 45/46

now I played with some combinations and ended in using the following code

```cpp
            case FILETYPE_FOLDER:
                size = std::min(size - 2, (u64)(0x10000 - addr));
                mem.poke(45, (addr + size) & 0xff );
                mem.poke(46, (addr + size) >> 8 );
                file.copyItem(nr, mem.ram + addr, size, 2);
                break;
```

which now results in the same runtime values of 45/46 compared the non flashing version

<img width="994" alt="image" src="https://github.com/vc64web/virtualc64web/assets/17108995/3cd1acff-9c2c-4c00-8856-ddf451c96ac6">

@dirkwhoffmann can you please review whether that makes sense ? Especially because I don't need to add 1 now...

**Comment by dirkwhoffmann** [comment]
It's correct if (and only if) it matches what we see when loading the program via the D64 route. Hence, comparing with the D64 version is the right way to go. Good work!

**Comment by mithrendal** [comment]
Ok fine … pushed the build to https://vc64web.github.io
